akka-cluster-sharding-typed/src/test/scala/docs/akka/cluster/sharding/typed/AccountExampleWithEventHandlersInState.scala

```diff
@@ -96,8 +98,8 @@ object AccountExampleWithEventHandlersInState {
     // to generate the stub with types for the command and event handlers.
 
     //#withEnforcedReplies
-    def apply(accountNumber: String): Behavior[Command[_]] = {
-      EventSourcedEntity.withEnforcedReplies(TypeKey, accountNumber, EmptyAccount, commandHandler, eventHandler)
+    def apply(entityContext: EntityContext): EventSourcedBehavior[Command[_], Event, Account] = {
+      EventSourcedBehavior.withEnforcedReplies(entityContext.persistenceId, EmptyAccount, commandHandler, eventHandler)
```

Comment:
EventSourcedBehavior must have the PersistenceId so the question is how to construct it. Here is a suggestion that the EntityContext provides the "standard" PersistenceId, but it must still be passed in as a parameter by the user to the EventSourcedBehavior. Would that be ok?

I was pondering if we could make the persistenceId optional so that it could be added after construction (by the EntityFactory in Play), but that doesn't work because it should also be possible to wrap the EventSourcedBehavior in other behaviors, such as `Behaviors.withTimers` and `Behaviors.supervise`, and then the EventSourcedBehavior type is lost (it's only a Behavior).

Reply:
> use PersistenceId(ctx.entityId)

yup, that is confusing. We should make it clear that the entityId is the business domain identity, as is. PersistenceId has the additional requirement to be globally unique, hence prefixing with the type name.

It's also a typical requirement that the entity itself has to know it's entityId, and therefore it can pass it from the EntityContext to the constructor.

> What I dislike in this approach is that if I'm not using an EventSourcedBehavior, I will still see a persistenceId method on my EntityContext. People will ask themselves why this exists? The same goes for EntityTypeKey.persistenceIdFrom though, so the problem is already there.

That is true, but I think we need to be somewhat pragmatic about that since we want to support the combination of ClusterSharding and EventSourcedBehavior as seamless as possible. That is the most common way to use ClusterSharding. We have to document that persistenceId can be ignored if Akka Persistence isn't used. entityId is always valid.

> If I'm a previous Akka untyped user, I will need to call it with PersistenceId(entityContext.entityId). But then I will see two methods on my EntityContext (entityId and persistenceId), and I will ask myself, why can't I simply use the persistenceId one?

I don't understand this part. `PersistenceId(entityContext.entityId)` would always be wrong.
If you have a previous persistenceId format it might be compatible with the EntityTypeKey format, with custom separator. Otherwise you have to construct it yourself from a plain `PersistenceId`, and can't use the EntityContext.persistenceId.

Maybe it should be named `EntityContext.defaultPersistenceId`.

Reply:
> I don't understand this part. `PersistenceId(entityContext.entityId)` would always be wrong.

It won't be wrong with I have been using Akka Persistence (untyped) and my journal is filled with rows using the business domain id. 

My concern is that we can confuse users with a method `persistenceId`. Such a user should not use that method, but `PersistenceId(entityContext.entityId)`. 

I don't think that `defaultPersistenceId` is conveying enough meaning. That's why I was going on the direction of a method that explicitly says: "here you get you business domain id with a type hint to make it globally unique"

Reply:
> It won't be wrong if I have been using Akka Persistence (untyped) and my journal is filled with rows using the business domain id.

sure, but then you have selected the persistenceId badly. Orange with "123" must not have the same persistenceId as Apple "123".